### DISCUSSION TITLE
$extends with needs: {relation: ?}

### DISCUSSION THREAD

**Discussion by rafma0** [prisma/prisma#20027] [Q&A] [answered]
Hi, I'm trying to create a result extension to make product names based on multiple relations, something like

```ts
  result: {
    product: {
      name: {
        needs: { nature: true, volume: true },
        compute (product) {
          if (!product.nature) return undefined
          return `${product.nature.name} (${product.volume})`
        }
      }
    }
  }
```

it works fine and I even get autocompletion for `product.nature`, problem is ts complains about 'nature' inside 'needs' with `not assignable to type 'never'`

not sure its a bug, I couldn't find any examples of using extensions with relations so I'm posting here first, ty

**Comment by rafma0** [comment]
What I'm doing for now is removing `needs` and manually typing `product` with relations and subrelations (ex: `product.package.color.name`) I want to use, etc

**Reply by nurul3101** [reply]
Hey @rafma0 👋 
Do you mind sharing your schema file so that I can try to reproduce this locally and check?

**Reply by rafma0** [reply]
Hi @nurul3101, I extract the relevant part from my project and put a minimal reproduction here https://github.com/rafma0/prisma-extends , please give it a look 😄

**Reply by nurul3101** [reply]
Hey @rafma0 👋 
Thanks for sharing the repository. I was able to run it locally and understand your use case now.
You are getting this type error for `nature` and `package` fields because they are relation fields and not scalar fields.

Only scalar fields are allowed in the `needs` parameter. From [Docs](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/result#considerations-for-fields):
> You can only create computed fields that are based on scalar fields

So this is expected. 
Also, here are various [examples](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions/extension-examples) that you can use for a reference.
<img width="795" alt="Screenshot 2023-07-04 at 3 55 05 PM" src="https://github.com/prisma/prisma/assets/13049676/7adb46a9-ad40-4fde-8818-39e214b3d394">

**Reply by rafma0** [reply]
@nurul3101 I see what you mean but even if the documentation says I can only create computed fields based on scalar fields, if I ignore the `needs` param I ***can*** create them from relation fields. It works. (you probably saw it when you ran my code first, without changes).

I can get values from relations inside the `compute` if I include the relations in the query, and I can make the computed field "optional" if I return `undefined` for cases where the relation does not exist (or are not included).

If we consider it already works, even if `needs` is not aware of it, and its a legit use case that makes compute fields more useful, wouldn't it be just a matter of doing some typescript "magic" to make `needs` aware of relations?

**Reply by nurul3101** [reply, marked as answer]
> you probably saw it when you ran my code first

Yes I did, and it indeed works!

We shared this case with our engineering team for their input and can confirm that we do plan to support using relation fields in the need clause. For now, you can use this workaround, and it should work.

Also, It would be great if you could **mark this Discussion as answered** to indicate that it has been resolved.

**Reply by nikolasburk** [reply]
Hey @rafma0 👋 just wanted to let you know that we now have an open issue for this that tracks support for relation fields in `result` extensions without the need to use `include`: https://github.com/prisma/prisma/issues/20091
